File: programmers/GRAPH/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

constexpr int MX = 20000;

bool relation[MX + 1][MX + 1];

struct node{
    int idx, d;
};

void makeRelationArr(int n, const vector<vector<int>> & edge){
    for(int i = 1; i <= n; ++i){
        for(int j = 1; j <= n ; ++j){
            relation[i][j] = false;
        }
    }
    
    int node1, node2;
    for(int idx = 0; idx < edge.size(); ++idx){
        node1 = edge[idx][0];
        node2 = edge[idx][1];
        relation[node1][node2] = relation[node2][node1] = true;        
    }
}

int searchLongestCount(int n){
    vector<bool> visited(n + 1);
    
    queue<node> q;
    
    q.push({1, 0});
    visited[1] = true;
    
    int longest = -1;
    int longestCnt = 0;
    while(!q.empty()){
        node currentNode = q.front();
        q.pop();
        
        int curIdx = currentNode.idx;
        int curD = currentNode.d;
        
        if(longest == curD){
            ++longestCnt;
        }
        else if(longest < curD){
            longest = curD;
            longestCnt = 1;
        }
        
        for(int nextIdx = 1; nextIdx <= n; ++nextIdx){
            if(!relation[curIdx][nextIdx])  continue;
            if(visited[nextIdx])    continue;
            visited[nextIdx] = true;
            q.push({nextIdx, curD + 1});
        }
    }
    return longestCnt;
}

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    makeRelationArr(n, edge);
    answer = searchLongestCount(n);
    return answer;
}

File: programmers/GRAPH/가장_먼_노드.java
package com.kt.aicare.common.util.programmers.GRAPH;
import java.util.*;

public class 가장_먼_노드 {

  final int NODE = 0;
  final int LENG = 1;
  final int EMPTY = -1;
  int nodeCnt = 0;
  List<Set<Integer>> relation = new ArrayList<>();
  List<Integer> visited = new ArrayList<>();

  void initData(int n, int[][] edge) {
    nodeCnt = n;
    relation = new ArrayList<>();
    visited = new ArrayList<>();
    for (int i = 0; i < nodeCnt + 1; i++) {
      relation.add(new HashSet<>());
      visited.add(EMPTY);
    }
    for (int[] e : edge) {
      int nodeA = e[0];
      int nodeB = e[1];
      relation.get(nodeA).add(nodeB);
      relation.get(nodeB).add(nodeA);
    }
  }

  void getResult() {
    Deque<int[]> paths = new LinkedList<>();
    Map<Integer, Integer> cache = new HashMap<>();
    paths.offerLast(new int[]{1, 0});
    visited.set(1, 0);
    while(!paths.isEmpty()) {
      int[] current = paths.pollFirst();
      for (Integer nextNode : relation.get(current[NODE])) {
        if (visited.get(nextNode) != EMPTY)
          continue;
        paths.offerLast(new int[]{nextNode, current[LENG] + 1});
        visited.set(nextNode, current[LENG] + 1);
      }
    }
  }

  public int solution(int n, int[][] edge) {
    initData(n, edge);
    getResult();
    int answer = 0;
    int mx = 0;
    for (int v : visited) {
      mx = Math.max(mx, v);
    }

    for (int v : visited) {
      if (v == mx) {
        answer++;
      }
    }


    return answer;
  }
}
